### PR TITLE
Shard API (Mark shard(s) as read only)

### LINKED ISSUE #1839: Mark shard(s) as read only
## Goals

* [ ] Introduce minimal API to view the status of all shards and update the status of a shard
* [ ] Track the amount of write requests and status
* [ ] Use minimal locking (more details in video)

## Notes

For additional context, please see [this video on the internal issue tracker](https://semi-technology.atlassian.net/browse/WEAVIATE-4).

### Get status of current shards
```
GET /v1/schema/{ClassName}/shards/

Example: 

GET /v1/schema/Author/shards/
Header: Content-Type: application/json

[{
  "name": "shard_12345",
  "status": "READY",
},{
  "name": "shard_abcdef",
  "status": "READONLY",
}]
```

### Update Status of shard

The following example would take an existing shard mark it read-only
```
PUT /v1/schema/{ClassName}/shards/{shardName}

Example: 

PUT /v1/schema/Author/shards/shard_12345
Header: Content-Type: application/json

{
  "status": "READONLY",
}
```

### PR DESCRIPTION
**Pull request by parkerduckworth**
## Changes 

Introduce shard API, with endpoints:

- `GET /v1/schema/{className}/shards`
- `PUT /v1/schema/{className}/shards/{shardName}`

These endpoints facilitate marking a shard as `READONLY`, which is useful in situations where the shard must be immutable; taking a disk snapshot for example.

Under the hood, if a shard is marked `READONLY`, segment compaction and memtable flushing are halted until the shard is reset to `READY`. In this case, Weaviate will emit logs indicating that a compaction would have occurred, but was skipped due to the current shard status.

---

Fixes #1839

### CODE REVIEW

File: adapters/repos/db/lsmkv/bucket.go [hunk 1]
```diff
@@ -45,6 +46,9 @@ type Bucket struct {
 	legacyMapSortingBeforeCompaction bool
 
 	stopFlushCycle chan struct{}
+
+	status     storagestate.Status
+	statusLock *sync.Mutex
```
Comment (antas-marcin) [inline adapters/repos/db/lsmkv/bucket.go:51]: is there a reason why `sync.Mutex` is a pointer?

Comment (parkerduckworth) [reply adapters/repos/db/lsmkv/bucket.go:51]: I took the example from [the attached internal issue video](https://semi-technology.atlassian.net/browse/WEAVIATE-4), which declared this field as a pointer. I don't have a reason aside from that 🙂 can change to value if preferred

File: adapters/repos/db/lsmkv/segment_group.go [hunk 2]
```diff
@@ -41,6 +42,9 @@ type SegmentGroup struct {
 	// for backward-compatibility with states where the disk state for maps was
 	// not guaranteed to be sorted yet
 	mapRequiresSorting bool
+
+	status     storagestate.Status
+	statusLock *sync.Mutex
```
[the remark this answers is not in the document]
Comment (parkerduckworth) [reply adapters/repos/db/lsmkv/segment_group.go:47]: See [above](https://github.com/semi-technologies/weaviate/pull/1860#discussion_r827163073)

File: test/integration/run.sh [hunk 3]
```diff
@@ -25,5 +25,5 @@ else
   echo "Found no --include-slow flag, skipping the slow ones"
 fi
 
-go test -count 1 -coverpkg=./adapters/repos/... -coverprofile=coverage-integration.txt -race -tags=$tags "$@" ./adapters/repos/...
-go test -count 1 -race -tags=$tags "$@" ./usecases/classification/...
+go test -v -count 1 -coverpkg=./adapters/repos/... -coverprofile=coverage-integration.txt -race -tags=$tags "$@" ./adapters/repos/...
+go test -v -count 1 -race -tags=$tags "$@" ./usecases/classification/...
```
Comment (antas-marcin) [inline test/integration/run.sh:29]: I think we should remove `-v` or better maybe introduce a param for example `--verbose` that would add this `-v` if needed

**Pull request merged**